pytorch_lightning/trainer/trainer.py

```diff
@@ -1109,6 +1112,24 @@ def __attach_dataloaders(self, model, train_dataloader=None, val_dataloaders=Non
         if test_dataloaders is not None:
             model.test_dataloader = _PatchDataLoader(test_dataloaders)
 
+    def __attach_datamodule(self, model, datamodule=None):
+        
+        # We use datamodule if it's been provided on .fit or .test, otherwise we check model for it
+        datamodule = datamodule or getattr(model, 'datamodule', None)
+
+        # If we have a datamodule, attach necessary hooks + dataloaders
+        if datamodule:
+            if self.is_overridden('setup', datamodule):
```

Comment:
rather define this in `setup` of module
```
def setup(...):
    if getattr(self, 'datamodule', None):
        self.datamodule.setup(...)
```

Reply:
Better yet, let's have DataModule do its own thing. Let's not mess with muddying up Trainer logic. 

We can encourage users to pass their subclass of `DataModule` directly to fit/test after running preparation on rank 0. Hopefully this makes things easier to manage 😄 . 

```python
if __name__ == '__main__':

    # Init your datamodule
    dm = MNISTDataModule()

    # Run all  preparation including downloading, transforms, etc.
    dm.prepare_data()

    # Get anything you need to init model, such as input dims
    input_dims = dm.size(1) * dm.size(2)

    # Init model
    model = CoolSystem(input_dims=input_dims)

    # Init Trainer
    trainer = pl.Trainer(max_epochs=3)

    # Train on your datamodule's data
    trainer.fit(model, datamodule=dm)
```

Reply:
so you cannot pass datamodule to `CoolSystem`?